0015-authority.md

```diff
@@ -0,0 +1,478 @@
+
+This system is very limited and not really `authority` it is more what we have in MLAPI with `ownership`. It looks like the MLAPI `ownership` model has been inspired by the UNet `authority` model. The MLAPI `authority` model will improve MLAPI and allow users to do things which were not possible in UNet.
+
+## DOTS Netcode
+DOTS Netcode has a ghost mode which can be interpolated or predicted. The interpolated mode matches our 'server' mode. There is no mode for client authority. The predicted mode is something we do not support yet but can in the future.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Scrapping per NetworkBehaviour Authority
+See appendix for details. There are very little actual use cases for this and scrapping this will more then half the code size of the authority feature and lower the bandwidth usage of MLAPI.
+
+## Naming
+We might want to consider a different name for authority. Some candidates would be `AuthorityMode`, `ControlMode` or `GhostMode`(to match DOTS). Similar `HasAuthority` might be `HasControl`.
+
+## Snapshots / AOI
+There are a lot of unknowns on how our snapshot and AOI system will work with `NetworkVariablePermission.Authority`. We do not plan to resolve this unresolved question in the authority RFC as it only introduces an additional `NetworkVariablePermission`. The snapshot / AOI RFCs will answer this question.
+
+## AuthorityChanged Event
+Do we need an AuthorityChanged event? We already have `OnAuthorityGained` and `OnAuthorityLost` events but would we need an additional to event to signal if the authority mode itself changed?
+
+## Default NetworkVariable Permissions
+
+The default write permission for a NetworkVariable is currently `NetworkVariablePermission.ServerOnly` and the default write permission is `NetworkVariablePermission.Everyone`. We could consider changing the default write permission to `NetworkVariablePermission.Authority`. It is not clear whether that would simplify or complicate the onboarding experience of MLAPI.
+
+## Shared/Distributed Authority
+
+The RFC does not cover authority modes where multiple peers have authority over an object. It also does not cover prediction. The proposed authority system can be extended to support more authority modes. To support shared/distributed authority we might need an ACL entry to indicate which peers have authority over an object
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Prediction
+The authority feature is a first step towards the goal of supporting client side prediction. When using client side prediction we need precises control over what's being run on the predicted client and run the same on the server. The `HasAuthority` flag will allow us to do this by setting it to true on both the predicting client and the server.
+
+## Commands
```

Comment:
I can't really see too much difference between commands vs RPCs. It feels like a more dedicated & specialized RPC implementation which brings me to the fact that we could simply extend and leverage existing RPC infrastructure.
Also, another note, both `NetVar` and `RPC` are just transport utilities, there is nothing wrong with carrying input data over an RPC and changing/simulating state on the server-side, then replicating server-sealed state back to all clients. There are no technical limits implementing input/predict/simulate/correct mechanism in the combination of `ServerRPC`s + pos/rot/... `NetVar`s.

Reply:
Command is more of a concept and not a specific implementation.
- RPCs can be commands
- NetworkVariable (state) can be commands.

Both have advantages/disadvantages in some situations. State can be rolled back, delta compressed and stored in the snapshot system. RPCs are much more flexible for one time commands. Most games get much better result by using state based commands for regular gameplay. DOTS Netcode also supports both ideas: <https://docs.unity3d.com/Packages/com.unity.netcode@0.0/manual/command-stream.html>

There are no hard technical limits which prevent RPCs commands to be predicted but we would need to introduce significant changes to RPCs. NetworkVariables do not need those changes.